cilium/CFP-27752-operator-manages-cilium-identities.md

```diff
@@ -0,0 +1,225 @@
+
+CIDs are presenting issues with reliability and scalability, and are causing network policies to be in a broken state. Most notable cases:
+
+* Reaching maximum number (65k) of CIDs in the cluster.
+* eBPF policy maps overflowing with over 16k entries.
+
+# Design
+
+## Operator changes
+
+Operator will start managing CIDs, and will become the only system pod that issues mutating requests for CIDs to the kube-apiserver. A new controller (CID controller) will run inside Operator that will be responsible for managing (creating) CIDs.
+
+1) Operator requires watching the following resources to allocate CIDs:
+
+* Pods (Already watches) - To manage CIDs based on pod labels.
+* Namespaces - to manage CIDs based on namespace labels changes.
+* CIDs - to keep the desired state of CIDs. Required for:
+    * Getting the initial state on start-up.
+    * Being compatible with Agents managing CIDs simultaneously.
+    * Handling unplanned CID changes (create, update, delete) by other sources -- not Operator or Agent.
+
+2) Operator will modify the following resources:
+
+* CIDs - Create and update. If namespace or pod labels are changed, new CIDs will be created for all unique label sets that don’t match any existing identity. Unused CIDs will be cleaned up by the CID GC. In case of an update to a CID, the CID controller will update the CID back to the correct state, effectively reverting the initial change.
+
+3) CID GC:
+
+* CID garbage collection continues to be the responsibility of the Operator. In the initial release, this will still rely on the legacy CID GC (the periodic GC) to deleted unused CIDs. The design of this can be evaluated after completing this CFP.
+
+## Agent changes
+
+Agents will stop creating CIDs. Agents will only generate local security identities. Agents will no longer perform following actions:
+
+* Locally create and reserve security identities in the global range [256, 65536].
+* Create CIDs.
+
+Subsequent milestones will further remove the following from the Agents:
+* Reporting heartbeat status annotation on CIDs.
+* Permissions to modify CIDs.
+
+Agents will wait for CIDs to be received through the watcher, and then assign them to local endpoints (pods) by retrieving the CIDs from the watcher's store based on pod and namespace labels.
+
+**Namespace watch**
+
+Agents need to know what are the namespace labels for each pod, to be able to match pod+namespace labels to a CID from CID watcher's cache. Agents will continue to watch namespaces in order to instantly update security identity labels in the local Endpoint manager’s cache, and start attempting to get CID from CID watcher store, to update the corresponding endpoint. More info below in Label changes: Namespace.
```

Comment:
> Agents will continue to watch namespaces in order to instantly update security identity labels in the local Endpoint manager’s cache, and start attempting to get CID from CID watcher store, to update the corresponding endpoint

Why is this necessary? Isn't the agent already listening for CIDs? And won't the new CID be a different CID anyway? How would the agent know which identity number to look for?

Reply:
I guess the answer is that the agent receives the namespace change signal to identify which endpoints must be updated, then logically it can put those endpoints into a "waiting-for-identity" mode which then needs to be correlated with a subsequent CID creation event to actually assign the new identity to the endpoint and close the loop. :+1:

Reply:
> start attempting to get CID from CID watcher store

This part seems like a no-op to me. The agent should already be listening for CIDs, and it wouldn't know which identity number to look for. There just needs to be a new callback on CID update events when there are endpoints looking for an updated identity, then it needs to match the updated identity with the new desired labels for the endpoint and only update the endpoint/CEP if there is a new identity that matches the pod+namespace labels.

Reply:
This paragraph just states how it will work, not necessarily changing the namespace watcher or how labels are updated in local Endpoints. Basically, when namespace labels change, everything will be the same, except at the end where it needs to create a CID and assign it, instead it will wait for matching CID based on labels to be present in the watcher's cache. This is part of the change to identity allocation for new pods.